src/mintpy/los2slope.py

```diff
@@ -0,0 +1,373 @@
+    geotransform = (
+        float(meta["X_FIRST"]),
+        float(meta["X_STEP"]),
+        0,
+        float(meta["Y_FIRST"]),
+        0,
+        float(meta["Y_STEP"]),
+    )
+    dataset = driver.Create("dem.tif", rows, cols, 1, gdal.GDT_Float32)
+    dataset.SetGeoTransform(geotransform)
+    dataset.SetProjection("EPSG:4326")
+    band = dataset.GetRasterBand(1)
+    band.WriteArray(dem_array)
+
+    # Calculate slope, save it as slope.tif as DEMProcessing does not allow mem rasters
+    slope_ds = gdal.DEMProcessing("slope.tif", dataset, "slope", scale=111120)
+    slope_array = slope_ds.ReadAsArray()
+
+    # Calculate aspect save it as aspect.tif as DEMProcessing does not allow mem rasters
+    aspect_ds = gdal.DEMProcessing(
+        "aspect.tif",
+        dataset,
+        "aspect",
+        zeroForFlat=True,
+    )
+    aspect_array = aspect_ds.ReadAsArray()
+
+    # Close all rasters
+    slope_ds = None
+    aspect_ds = None
+    dataset = None
+    return slope_array, aspect_array
+
+
+def get_design_matrix4slope(los_inc_angle, los_az_angle, slope, aspect):
+    """Design matrix G to convert asc/desc range displacement into slope direction.
+    Only 2D approach is implemented.
+
+    Project displacement from LOS to slope direction components:
+
+    Math for 2D:
+        dLOS =   dslope * G
+        where, G = (sin(inc_angle) * cos(az_angle - aspect) * cos(slope) + cos(inc_angle) * sin(slope))
+
+    Parameters: los_inc_angle - 1/2D np.ndarray in size of (num_file, window_y, window_x), LOS incidence angle in degree.
+                los_az_angle  - 1/2D np.ndarray in size of (num_file, window_y, window_x), LOS azimuth angle in degree, measured in anti-clockwise direction from east.
+                slope         - 1/2D np.ndarray, slope of the terrain in degree.
+                aspect        - 1/2D np.ndarray, aspect of the terrain in degree, measured in anti-clockwise direction from east.
+    Returns:    G             - 2D matrix
+    """
+    # 2D approach: not currently working, needs math re-check
+    # aspect = -aspect
+    # vh = np.sin(np.deg2rad(los_inc_angle))*np.cos(np.deg2rad(los_az_angle-aspect))
+    # vu = np.cos(np.deg2rad(los_inc_angle))
+    # ch = np.cos(np.deg2rad(slope))
+    # cu = np.sin(np.deg2rad(slope))
+    # G = vh*ch + vu*cu
+
+    # # 3d Approach
+    aspect = (
+        -aspect + 90
+    )  # Shift hillsope aspect to 0 east anti-clockwise, same as the azimuth angle.
+
+    ve = np.sin(np.deg2rad(los_inc_angle)) * np.sin(np.deg2rad(los_az_angle)) * -1
+    vn = np.sin(np.deg2rad(los_inc_angle)) * np.cos(np.deg2rad(los_az_angle))
+    vu = np.cos(np.deg2rad(los_inc_angle))
+
+    ce = (np.sin(np.deg2rad(aspect)) * np.sin(np.deg2rad(slope))) * -1
+    cn = np.sin(np.deg2rad(aspect)) * np.cos(np.deg2rad(slope))
+    cu = np.cos(np.deg2rad(slope))
+
+    G = ve * ce + vu * cu + vn * cn
+    return G
+
+
+def los2slope(dlos, los_inc_angle, los_az_angle, slope, aspect, inps, step=20):
```

Comment:
suggest to replace _inps_ with the input parameters:
slp_thresh=5.0, asp_thresh=0.0, scaling_factor_thresh=10, step=20

to make this function modular

Reply:
I second this suggestion